Add a database for samples

sakertooth opened the pull request:
Adds a database for samples. Callers can fetch data from the database using two overloads: one for audio files, and another for Base64 strings. The database stores weak pointers to the samples and returns shared pointers to the callers. However, the database in the future will probably need to store the weak pointers along with sample thumbnails and possibly other kinds of metadata for each sample, which should be fairly easy to do as all that needs to be done is to store a collection of the data we need in a `struct`/`class` instead of just the weak pointer to the buffer.

For audio files, we check the last write time for the file to see if it needs updating each time it is fetched, and update it if necessary. For Base64 strings, we just compare the contents.

`SampleLoader` loads audio data by querying the database for it, rather than creating them manually. As such, the `SampleLoader::create*` functions where renamed to `SampleLoader::load*`.

The memory usage has dropped significantly when using duplicate samples (`htop` readings went from 28.5% to 3.6% for me with a project that had a number of sample clips, each ~3 minutes long), and projects load faster (There is still some delay because of the waveform drawing, but this should be addressed soon. The speedup is more noticeable when you zoom all the way out before loading a project as a result).

Should supersede #7058 I believe.

PS: This is sample caching, but branded under being a sample database instead of a sample cache. Since this PR uses two hashmaps (which can resemble tables in a database), a database seems like a more accurate term, but if there are objections we can stick to saying it is a sample cache.

messmerd commented on include/SampleLoader.h:
`SampleLoader` should be in the `lmms` namespace, not `lmms::gui`.

If you want, you could move the open dialog methods to a separate class under the the `lmms::gui` namespace like I did here:
[include/SampleLoaderDialog.h](https://github.com/LMMS/lmms/blob/af540e9bb76734257b22a99760f3931be1ac5ee4/include/SampleLoaderDialog.h)

sakertooth replied:
Why? The `load*` functions need to access the GUI to display errors. There are really just wrappers to handle errors with buffers that failed to load. Otherwise, we wouldn't need them at all and callers could just use `SampleDatabase::fetch` directly. Unless the burden of checking if a buffer failed to load should be on the callers?

messmerd replied:
I think we had this conversation before. `SampleLoader` has nothing to do with the GUI, so it shouldn't be in the `gui` namespace. When rendering projects via command-line, LMMS still needs to use this class.

sakertooth replied:
> I think we had this conversation before. SampleLoader has nothing to do with the GUI, so it shouldn't be in the gui namespace.

What makes `SampleLoader` not related to the GUI? Is it because the name doesn't give off a feeling that it belongs there? As you mentioned it has the `open*` functions that directly utilize the GUI, and the "loading" part of it is really just calling `SampleDatabase::fetch` directly. The only reason why the "loading" part is in the GUI is because it accesses the GUI to display errors so callers don't always have to repeat the same code that checks for failures to load.

I think the proper way to solve this is to move `SampleLoader` into the core, but emit a Qt signal so the GUI can respond accordingly.

messmerd replied:
There's a lot of `lmms::` namespace code that calls `gui::getGUI()` to check whether LMMS is running with a GUI or not. Displaying error messages to a user is one example where this is used, since with a GUI you can display a message box but without a GUI you need to print the error to the terminal.

If an error occurs when a project is being loaded, you should probably call
```c++
Engine::getSong()->collectError(...)
```

This method collects error messages, and later on, in `Song::loadProject` the error message is displayed:
```c++
if ( hasErrors())
	{
		if ( getGUI() != nullptr )
		{
			QMessageBox::warning( nullptr, tr("LMMS Error report"), errorSummary(),
							QMessageBox::Ok );
		}
		else
		{
#if (QT_VERSION >= QT_VERSION_CHECK(5,15,0))
			QTextStream(stderr) << Engine::getSong()->errorSummary() << Qt::endl;
#else
			QTextStream(stderr) << Engine::getSong()->errorSummary() << endl;
#endif
		}
	}
```

Even though `Song` conditionally interacts with the GUI, it is not in `lmms::gui::`. That's because `Song` is a core part of LMMS which is needed even when running without a GUI. Same goes for `SampleLoader`.

sakertooth replied:
I'm actually just going to remove the `load*` functions in `SampleLoader`, change the class name so that it better matches the what the `open*` functions do, and make the callers call `SampleDatabase::fetch` directly and move any Qt signal functionality in there.

messmerd replied:
I would recommend moving the GUI-only dialog methods to a separate class `SampleLoaderDialog`.

sakertooth replied:
> Even though Song conditionally interacts with the GUI, it is not in lmms::gui::. That's because Song is a core part of LMMS which is needed even when running without a GUI. Same goes for SampleLoader.

That's fair, though we should probably move away from calling `getGUI` directly in the core at some point.

messmerd reviewed (commented):
Gave it a quick look-over.

Also,
- `Sample`'s move constructor and move assignment operator should be marked `noexcept`
- `Sample::s_interpolationMargins` should be renamed to `Sample::InterpolationMargins` since it is public
- `sample_rate_t` could be used instead of `int` for the sample rate (See this commit for what needed to be changed: https://github.com/LMMS/lmms/pull/7058/commits/10162ecd251cf32fdbe2607616ebfe4dbfc35083)

messmerd commented on include/SampleDatabase.h:
```diff
@@ -0,0 +1,125 @@
+		return entryLock;
+	}
+
+	inline static std::unordered_map<AudioFileEntry, std::weak_ptr<SampleBuffer>, Hash> s_audioFileMap;
+	inline static std::unordered_map<Base64Entry, std::weak_ptr<SampleBuffer>, Hash> s_base64Map;
```
You may need a mutex to guard access to these in case two threads try to load a sample concurrently

sakertooth replied:
Are there other threads besides the main thread that will fetch from `SampleDatabase`?

messmerd replied:
I'm not sure, but if you want to make sure only the main thread is calling this, you could use:

```c++
assert(QThread::currentThread() == QCoreApplication::instance()->thread());
```

sakertooth commented:
> Sample's move constructor and move assignment operator should be marked noexcept
    Sample::s_interpolationMargins should be renamed to Sample::InterpolationMargins since it is public
    sample_rate_t could be used instead of int for the sample rate (See this commit for what needed to be changed: https://github.com/LMMS/lmms/commit/10162ecd251cf32fdbe2607616ebfe4dbfc35083)

Thanks, I will make another PR to address these issues separately. However, about the `int` -> `sample_rate_t` change, I never really understood why we have those types when basic types like `int` will work just as well. We already have to convert the sample rate to an `int` when processing the project file anyways. I can't really see the benefit using `sample_rate_t` brings, and it adds more complexity than `int` does (not in terms how long the type is.. well maybe, but more so having to know what is the actual type is that we are dealing with here).

sakertooth commented:
Any updates @messmerd?

messmerd commented on include/SampleCache.h:
```diff
@@ -0,0 +1,125 @@
+	};
+
+	template <typename T, typename ...Args>
+	static auto get(const T& entry, std::unordered_map<T, std::weak_ptr<SampleBuffer>, Hash>& map, Args... args)
+	{
+		const auto it = map.find(entry);
+
+		if (it == map.end())
+		{
+			const auto buffer = std::make_shared<SampleBuffer>(std::forward<Args>(args)...);
+			map.insert(std::make_pair(entry, buffer));
+			return buffer;
+		}
+
+		const auto entryLock = it->second.lock();
+		if (!entryLock)
+		{
+			const auto buffer = std::make_shared<SampleBuffer>(std::forward<Args>(args)...);
+			map[entry] = buffer;
+			return buffer;
+		}
+
+		return entryLock;
+	}
```
This function makes a copy of `entry` when it could be moved into this function from the call sites to avoid copying the file path or base64 data.

If the first parameter was `T entry`, you could emplace into the maps
```cpp
map.emplace(std::move(entry), buffer);
```

messmerd commented:
This PR looks pretty good to me code-wise. I'll test it next.

messmerd commented on include/SampleCache.h:
```diff
@@ -0,0 +1,125 @@
+		const auto entryLock = it->second.lock();
+		if (!entryLock)
+		{
+			const auto buffer = std::make_shared<SampleBuffer>(std::forward<Args>(args)...);
+			map[entry] = buffer;
```
Here you should be able to reuse the `it` iterator since it already points to the location where you'll want to put the new buffer. No need for a 2nd lookup.

messmerd commented on include/SampleCache.h:
```diff
@@ -98,23 +98,16 @@ class SampleCache
 	static auto get(const T& entry, std::unordered_map<T, std::weak_ptr<SampleBuffer>, Hash>& map, Args... args)
 	{
 		const auto it = map.find(entry);
+		const auto item = it == map.end() ? nullptr : it->second.lock();
```
This last commit didn't fix either of the problems I mentioned. `entry` still has an extra copy and the map iterator is not reused so a lookup occurs twice (once in `find` and again in `emplace`).

I was thinking something like this:
```cpp
template <typename T, typename... Args>
static auto get(T entry, std::unordered_map<T, std::weak_ptr<SampleBuffer>, Hash>& map, Args&&... args)
{
	const auto it = map.find(entry);

	if (it == map.end())
	{
		const auto buffer = std::make_shared<SampleBuffer>(std::forward<Args>(args)...);
		map.emplace(std::move(entry), buffer);
		return buffer;
	}

	const auto entryLock = it->second.lock();
	if (!entryLock)
	{
		const auto buffer = std::make_shared<SampleBuffer>(std::forward<Args>(args)...);
		it->second = buffer;
		return buffer;
	}

	return entryLock;
}
```

Calling code (just `SampleCache::fetch` I think) should move `entry` when calling.

I also made the `args` parameter an rvalue reference because that is needed for perfect forwarding.

sakertooth replied:
Oops, my bad, fixing now.

sakertooth replied:
Wait, `emplace` does a lookup?

sakertooth replied:
I actually think we can use `insert_or_assign` here.

messmerd replied:
I don't think `insert_or_assign` would be useful in this situation. The "insert" part is good, but you don't want to unconditionally assign to the map since you need to call the `lock` method on the weak pointer and return the result without reassigning it if it is non-null.

messmerd replied:
The issues I mentioned still aren't solved. `insert_or_assign` still needs to look up the key, and you shouldn't call `std::move` on a const reference.

The code I shared in my previous comment fixes both. However, while it did fix the unnecessary lookup in the `!entryLock` branch, there was still a 2nd lookup in the `it == map.end()` branch.

But now I think I've fixed that as well, while also simplifying the code:

```cpp
// NOTE: `entry` is a value not a reference, and calling code can use `std::move`
template <typename T, typename... Args>
static auto get(T entry, std::unordered_map<T, std::weak_ptr<SampleBuffer>, Hash>& map, Args&&... args)
{
	auto& value = map[std::move(entry)]; // NOTE: This is the only lookup; moves from `entry` and default constructs value if key does not exist

	auto buffer = value.lock(); // NOTE: null if key didn't exist or previous buffer was freed
	if (!buffer)
	{
		buffer = std::make_shared<SampleBuffer>(std::forward<Args>(args)...);
		value = buffer;
	}

	return buffer;
}
```

The `NOTE` comments are added for explanation.

messmerd commented on src/core/SampleCache.cpp:
```diff
@@ -0,0 +1,53 @@
+	assert(QThread::currentThread() == QCoreApplication::instance()->thread());
+
+	const auto fsPath = PathUtil::pathFromQString(path);
+	const auto entry = AudioFileEntry{fsPath, std::filesystem::last_write_time(fsPath)};
+	return get(entry, s_audioFileMap, path);
```
```suggestion
	auto entry = AudioFileEntry{fsPath, std::filesystem::last_write_time(fsPath)};
	return get(std::move(entry), s_audioFileMap, path);
```
With the latest changes, this calling code can now move the entry to avoid the copy

sakertooth replied:
Great attention to detail, I commonly overlook things like this. Reminds me of how powerful yet deadly the language is, especially with things like using `auto` instead of `auto&` when storing a variable from a function returned by reference when you don't intend to make a copy (this is not related to the code here at all, just an example).

messmerd commented on src/core/SampleCache.cpp:
```diff
@@ -0,0 +1,53 @@
+	// Can be removed if proven that other threads are requesting samples besides the main thread
+	assert(QThread::currentThread() == QCoreApplication::instance()->thread());
+
+	const auto entry = Base64Entry{base64.toStdString(), sampleRate};
+	return get(entry, s_base64Map, base64, sampleRate);
```
```suggestion
	auto entry = Base64Entry{base64.toStdString(), sampleRate};
	return get(std::move(entry), s_base64Map, base64, sampleRate);
```
Same here

messmerd commented on src/core/SampleCache.cpp:
```diff
@@ -0,0 +1,53 @@
+	// Can be removed if proven that other threads are requesting samples besides the main thread
+	assert(QThread::currentThread() == QCoreApplication::instance()->thread());
+
+	const auto fsPath = PathUtil::pathFromQString(path);
```
```suggestion
	const auto fsPath = PathUtil::pathFromQString(PathUtil::toAbsolute(path));
```
`path` can be a relative path such as "usersample:Amen-break.wav", so I think that needs to be converted to an absolute path here. Otherwise `std::filesystem::last_write_time` will fail.

Spacemagehq commented:
I'm testing, and I'm finding no issues or bugs so far on Windows 11. I had 20 samples and samples randomly throughout lmms and the memory and the storage seems fine and not affected. The CPU meter inside lmms is not going up by that much and not heavy on the cpu.

sakertooth commented:
@messmerd, I believe I understand the benefit of using UUIDs better now. If the file path changes, then we only would need to update the mapping from UUIDs to file paths at one place, while all the other clients don't have to worry about anything and can continue using the UUID, which hasn't changed.

As of right now my PRs handles it using the "last modified time", which has two problems. One, this can create a lot of dead entries if they are not being actively "garbage collected". Two, since clients still store the file path, if it changes, they would have to still manage the sample by themselves, and that would have to be done possibly everywhere. This is a problem in #7366, where we have check if the `Thumbnail` needs to be updated or a new one to be created because we store the file paths directly, which can become invalidated at any point on the file system.

It seems like the overall idea here is to map UUIDs to assets/resources, which can be a sample file, sample Bas64 string, project file, preset file, etc. Assets can then specify if they are loaded from disk, or something else like as a string in the case of Base64 samples. Each asset can be updated as necessary to when the asset manager/cache feels like it should (changes on file system or something else possibly). This will not only bolster the sample caching implementation since it truly centralizes everything, but will also help make forward strides in simplifying and improving asset management (which may be needed for features like showing a popup to the user to load missing assets when loading the project, among other things).

I am going to try to implement some of these ideas in a new PR.

messmerd commented:
@sakertooth Yep, that's exactly it.

This PR doesn't make any changes to the project file as far as I'm aware, so it won't introduce any backwards incompatible changes if we merge it now. And since this PR is very useful as is, I think we should merge it then explore the UUID / resource manager idea in a follow-up PR.

sakertooth commented:
I'll at least move to using `QFileSystemWatcher` in the current implementation to fix some of the problems mentioned by actively keeping the table in sync with the file system. Other than that I think this is safe to merge. I agree that the UUID asset idea might need to be explored in a different PR since its implementation is far more lengthy than what I have here, and I remember you were planning to do this already to some extent.

sakertooth commented:
I read up on the docs for `QFileSystemWatcher`

> The act of monitoring files and directories for modifications consumes system resources. This implies there is a limit to the number of files and directories your process can monitor simultaneously. On all BSD variants, for example, an open file descriptor is required for each monitored file. Some system limits the number of open file descriptors to 256 by default. This means that [addPath](https://doc.qt.io/qt-6/qfilesystemwatcher.html#addPath)() and [addPaths](https://doc.qt.io/qt-6/qfilesystemwatcher.html#addPaths)() will fail if your process tries to add more than 256 files or directories to the file system monitor. Also note that your process may have other file descriptors open in addition to the ones for files being monitored, and these other open descriptors also count in the total. macOS uses a different backend and does not suffer from this issue.

I was a bit scared off making the switch because of this and had to weigh the pros and cons, so I'll probably leave the timestamp checking to be safe as it works well enough. A counter argument to having dead paths in the caches is that its rare and have static duration, so it really not that big of a problem as I made it out to be. The main issue is exposing the file path information that has to be made in sync everywhere else in the codebase, but we will deal with this later as already discussed.

One change I should make though is that instead of making completely new entries, update preexisting ones if they are fetched a second time and are still valid on the file system.

messmerd commented on src/core/SampleCache.cpp:
```diff
@@ -32,22 +32,43 @@
-	const auto fsPath = PathUtil::pathFromQString(PathUtil::toAbsolute(path));
-	auto entry = AudioFileEntry{fsPath, std::filesystem::last_write_time(fsPath)};
-	return get(std::move(entry), s_audioFileMap, path);
+	const auto it = std::find_if(s_audioFileMap.begin(), s_audioFileMap.end(),
+		[&](const auto& entry) { return entry.first.path == PathUtil::pathFromQString(path); });
+
+	auto lastWriteTime = fs::last_write_time(PathUtil::pathFromQString(path));
+
+	if (it == s_audioFileMap.end() || it->first.lastWriteTime != lastWriteTime)
+	{
+		const auto buffer = std::make_shared<SampleBuffer>(path);
+		const auto key = AudioFileEntry{PathUtil::pathFromQString(path), lastWriteTime};
+		s_audioFileMap[std::move(key)] = buffer;
+		return buffer;
+	}
```
```suggestion
	const auto qPath = PathUtil::pathFromQString(path);
	const auto it = std::find_if(s_audioFileMap.begin(), s_audioFileMap.end(),
		[&](const auto& entry) { return entry.first.path == qPath; });

	auto lastWriteTime = fs::last_write_time(qPath);

	if (it == s_audioFileMap.end() || it->first.lastWriteTime != lastWriteTime)
	{
		const auto buffer = std::make_shared<SampleBuffer>(path);
		const auto key = AudioFileEntry{qPath, lastWriteTime};
		s_audioFileMap[std::move(key)] = buffer;
		return buffer;
	}
```
Moved the `PathUtil::pathFromQString(path)` outside the lambda so it doesn't call it every time the lambda is called.

sakertooth replied:
I would use `fsPath` instead personally, but yeah I'll move it out.

messmerd commented on plugins/AudioFileProcessor/AudioFileProcessor.cpp:
```diff
@@ -195,9 +194,9 @@ void AudioFileProcessor::deleteNotePluginData( NotePlayHandle * _n )
+	if (!m_sample.sampleFile().isEmpty())
 	{
-		elem.setAttribute("sampledata", m_sample.toBase64());
+		elem.setAttribute("sampledata", QString::fromStdString(m_sample.buffer()->toBase64()));
```
The resource manager still needs significantly more work I think. You're still saving the base64 data for each AFP here rather than creating a new section of the project file dedicated to resource storage. While base64 data is cached in memory, it will not be "cached" on disk until it exists in a single place in the project file rather than potentially multiple.

That's the functionality the resource manager can provide, but to do that, it will need to perform the job of loading/saving resources to/from the project file itself. AFP and other users of the resource manager would only be responsible for loading/saving a resource UUID and calling the resource manager to load/save the resource. AFP should never deal with base64 or sample file paths except during a one-time upgrade from the current system to the UUID system in the `loadSettings` method. Once that upgrade routine is performed, AFP and other resource manager users should never need to touch a base64 string or sample file path again since that will be the resource manager's responsibility.

It's going to be a lot of work and will require changes to the project files, which is why I recommended adding the resource manager in a separate follow-up PR and letting this PR just add sample caching. Sample caching is an important optimization that has stalled for far too long, and you had a sample caching system that was about ready to merge until you started adding the resource manager stuff here.

So I'd recommend moving the recent resource manager additions (https://github.com/LMMS/lmms/pull/7497/commits/38433daa847970d6c5407e21a1d6b40ca7bfc5c8 and all commits after it) to a separate branch and letting this PR be the sample cache only. The sample cache exists purely at runtime and does not modify the project file like a proper resource manager implementation would, so it should not create any compatibility issues with the design of the future resource manager if it is merged now.

sakertooth replied:
With the implementation I had before, I believe I discovered some issues with it. However, what I will say is that this resource manager utilizes hashes (MD5 hashing with `QCryptographicHash`) and also stores the cached resources as shared pointers. It's been some time since I made these changes, but IIRC, the main issues were storing weak pointers to the cached entries, and storing Base64 audio data as a key.

Storing weak pointers is an idea I keep going back and fourth with, so I apologize if this is repetitive. The conflict I'm realizing with storing weak pointers is that as soon as a resource is deleted, it will also be erased from the cache immediately, thus preventing any performance benefit in situations where that resource will be needed again sometime shortly. Storing weak pointers works fine if we only are worrying about minimizing memory usage (since we keep track of all possible resources currently used at the moment), but I think the performance detriment I mentioned earlier can be quite extreme depending on the situation.

Storing Base64 audio data as a key is another problem. This can be far too much data, and its very similar to simply storing all the loaded samples in memory. It does not make much sense to cache them for this very reason. We already have the data, we just need to decode it and choose a sample rate for the samples to follow. The problem is that we can now have sample thumbnails as being cached and loaded from the Base64 data. So to prevent storing the Base64 audio data as a key in tandem with the sample thumbnail resource, it was hashed using MD5 and its digest stored in a string, which is a lot less data to deal with. I did something similar for resources that came from a file path by hashing its contents in chunks instead.

The rebranding of the `SampleCache` to `ResourceCache` happened in preparation for the general resource manager. This was another issue I was considering, because now we have to not only cache buffers of samples, but also sample thumbnails. The introduction of another resource to manage pushed me to future-proof a bit since I was certain we were going to cache multiple kinds of resources.

There were other variables as I was considering, but most of problems all revolved around how much of a "cache" the `SampleCache` really was. I was considering issues that were making or could've made the implementation at the time fairly poor. I probably still need time to go over this PR a couple more times as I work out the problem a bit more.

sakertooth replied:
I also wanted to point out that my implementation does not deal with the project files at all. I'm more focused on caching resources within memory, while caching resources in the project file by centralizing them can be done at a later point. That might require a different kind of class altogether since the responsibilities aren't so similar.

sakertooth replied:
Maybe this is overkill or too confusing, I'll try to work on a new, hopefully simpler solution.

sakertooth commented:
Hi @messmerd, I scrapped the PR again and implemented it similar to how I did it the first time but with a different approach: I made a new class (`FileCache`) that can cache any kind of type originating from a file. Currently, only `SampleBuffer` and `SampleThumbnail` store a `FileCache<SampleBuffer>` and `FileCache<SampleThumbnail>` respectively in the `private` section of their declaration.

Since `FileCache` uses `std::filesystem::path` and we eventually want to migrate to it, I figured I wouldn't convert to and from a `QString` within the core, so I have a few changes that span a couple of files to make them use `std::filesystem::path` instead. I also added new `PathUtil::fsConvert` overloads to convert to and from a `QString`/`std::filesystem::path`.

I did not cache Base64, but it can be added if necessary. Reason being is that Base64 is the data itself (as you know) and isn't really a key in the cache like a file path is. Since there really isn't any good key for that kind of data (except hashing it, which can be slow) as of right now, I have ignored it.

I don't plan on going against the grain from this point, apologies if there was confusion on my end about this. I suppose I made things more complicated than they had to be. This PR is basically done on my end.
